src/providers/tpm_provider/utils.rs

```diff
@@ -84,3 +91,134 @@ pub struct PasswordContext {
+            };
+            Ok(KeyParams::Rsa {
+                size,
+                scheme: convert_asym_scheme_to_tpm(attributes.key_policy.key_algorithm)?,
+                pub_exponent: 0,
+            })
+        }
+        KeyType::EccKeyPair { curve_family } => Ok(KeyParams::Ecc {
+            scheme: convert_asym_scheme_to_tpm(attributes.key_policy.key_algorithm)?,
+            curve: convert_curve_to_tpm(curve_family, attributes.key_bits)?,
+        }),
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+fn convert_asym_scheme_to_tpm(algorithm: Algorithm) -> Result<AsymSchemeUnion> {
+    match algorithm {
+        Algorithm::AsymmetricSignature(AsymmetricSignature::RsaPkcs1v15Sign { hash_alg }) => {
+            Ok(AsymSchemeUnion::RSASSA(convert_hash_to_tpm(hash_alg)?))
+        }
+        Algorithm::AsymmetricSignature(AsymmetricSignature::Ecdsa { hash_alg }) => {
+            Ok(AsymSchemeUnion::ECDSA(convert_hash_to_tpm(hash_alg)?))
+        }
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+#[allow(deprecated)]
+fn convert_hash_to_tpm(hash: Hash) -> Result<HashingAlgorithm> {
+    match hash {
+        Hash::Sha1 => Ok(HashingAlgorithm::Sha1),
+        Hash::Sha256 => Ok(HashingAlgorithm::Sha256),
+        Hash::Sha384 => Ok(HashingAlgorithm::Sha384),
+        Hash::Sha512 => Ok(HashingAlgorithm::Sha512),
+        Hash::Sha3_256 => Ok(HashingAlgorithm::Sha3_256),
+        Hash::Sha3_384 => Ok(HashingAlgorithm::Sha3_384),
+        Hash::Sha3_512 => Ok(HashingAlgorithm::Sha3_512),
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+fn convert_curve_to_tpm(curve_family: EccFamily, key_size: u32) -> Result<EllipticCurve> {
+    match (curve_family, key_size) {
+        (EccFamily::SecpR1, 192) => Ok(EllipticCurve::NistP192),
+        (EccFamily::SecpR1, 224) => Ok(EllipticCurve::NistP224),
+        (EccFamily::SecpR1, 256) => Ok(EllipticCurve::NistP256),
+        (EccFamily::SecpR1, 384) => Ok(EllipticCurve::NistP384),
+        (EccFamily::SecpR1, 512) => Ok(EllipticCurve::NistP521),
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+pub fn pub_key_to_bytes(pub_key: PublicKey, p_byte_size: usize) -> Result<Vec<u8>> {
+    match pub_key {
+        PublicKey::Rsa(key) => picky_asn1_der::to_vec(&RsaPublicKey {
+            modulus: IntegerAsn1::from_unsigned_bytes_be(key),
+            public_exponent: IntegerAsn1::from_signed_bytes_be(PUBLIC_EXPONENT.to_vec()),
+        })
+        .or(Err(ResponseStatus::PsaErrorGenericError)),
+        PublicKey::Ecc { x, y } => {
+            if x.len() != p_byte_size || y.len() != p_byte_size {
+                error!(
+                    "Received ECC public key with invalid size: x - {} bytes; y - {} bytes",
+                    x.len(),
+                    y.len()
+                );
+                return Err(ResponseStatus::PsaErrorGenericError);
```

Comment:
Isn't it `InvalidArgument` for this one?

Reply:
or `CommunicationFailure`, if the information comes from the TPM?

Reply:
Maybe `CommunicationFailure` - it isn't invalid argument because the public key comes from us, not from the user.

Reply:
`CommunicationFailure` sounds right then

Reply:
Dun!